src/lang.rs

```diff
@@ -1,295 +1,597 @@
 }
 
-impl FunctionInfo {
-    pub fn new(
-        name: String,
-        f: &flux::semantic::types::Function,
-        package_name: String,
-    ) -> Self {
-        FunctionInfo {
-            name,
-            package_name,
-            required_args: get_argument_names(&f.req),
-            optional_args: get_optional_argument_names(&f.opt),
-        }
+impl std::cmp::Ord for Function {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.name.cmp(&other.name)
+    }
+}
+
+impl std::cmp::PartialOrd for Function {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
     }
+}
+
+impl PartialEq for Function {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+    }
+}
+impl Eq for Function {}
 
-    pub fn signatures(&self) -> Vec<FunctionSignature> {
-        let mut result = vec![FunctionSignature {
-            name: self.name.clone(),
-            arguments: self.required_args.clone(),
-        }];
+impl Function {
+    /// Get signature information for a flux function.
+    pub fn signature_information(
+        &self,
+    ) -> Vec<lsp::SignatureInformation> {
+        let required: Vec<String> =
+            self.expr.req.keys().map(String::from).collect();
+        let optional: Vec<String> =
+            self.expr.opt.keys().map(String::from).collect();
+        let mut result = vec![required.clone()];
 
         let mut combos = vec![];
-        let length = self.optional_args.len();
+        let length = optional.len();
         for i in 1..length {
             let c: Vec<Vec<String>> =
-                combinations::Combinations::new(
-                    self.optional_args.clone(),
-                    i,
-                )
-                .collect();
+                combinations::Combinations::new(optional.clone(), i)
+                    .collect();
             combos.extend(c);
         }
-        combos.push(self.optional_args.clone());
+        combos.push(optional);
 
         for l in combos {
-            let mut arguments = self.required_args.clone();
+            let mut arguments = required.clone();
             arguments.extend(l.clone());
 
-            result.push(FunctionSignature {
-                name: self.name.clone(),
-                arguments,
-            });
+            result.push(arguments);
         }
 
         result
-    }
-}
-
-pub struct FunctionSignature {
-    pub name: String,
-    pub arguments: Vec<String>,
-}
-
-impl FunctionSignature {
-    pub fn create_signature(&self) -> String {
-        let args: String = self
-            .arguments
-            .iter()
-            .map(|x| format!("{}: ${}", x, x))
-            .collect::<Vec<String>>()
-            .join(" , ");
+            .into_iter()
+            .map(|arguments| lsp::SignatureInformation {
+                label: {
+                    let args = arguments
+                        .iter()
+                        .map(|x| format!("{}: ${}", x, x))
```

Comment:
This is [snippet syntax](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax), which is a feature we want to use, and is the core thing driving all of these refactors. This change here made it so that completions are _only_ ever happening through this codepath, because when I originally tried to do this and this codepath was stepping on my work, leaving me scratching my head about why my changes weren't taking effect.